### PR TITLE
Allows basic mobs to smash things in their way

### PR DESCRIPTION
**Pull request by ghost**
This makes the environment_smash var be used when trying to path to an enemy. Basic mobs will try to smash things in their way if the appropriate environment_smash level is available. This is ridiculously broken on simple animals so I re-wrote it a bit for basic mobs. (The var is treated both like an enum and a bitfield)

This also allows you to designate actions as non-replan-blocking. Which means the controller can make a new plan, even if this action is running. This allows actions like this which are more passively ran to exist.

### CODE REVIEW

File: code/modules/vehicles/mecha/mecha_defense.dm [hunk 1]
```diff
@@ -82,11 +82,29 @@
...
 		animal_damage = min(animal_damage, 20*user.environment_smash)
 		log_combat(user, src, "attacked")
 		attack_generic(user, animal_damage, user.melee_damage_type, MELEE, play_soundeffect)
 		return 1
 
+/obj/vehicle/sealed/mecha/attack_basic_mob(mob/living/basic/user, list/modifiers)
+	log_message("Attack by basic mob. Attacker - [user].", LOG_MECHA, color="red")
+	if(!user.melee_damage_upper && !user.obj_damage)
+		user.emote("custom", message = "[user.friendly_verb_continuous] [src].")
+		return 0
```
Comment (Fikou) [inline code/modules/vehicles/mecha/mecha_defense.dm:95]: false true, same for the rest of the proc

File: code/game/turfs/closed/walls.dm [hunk 2]
```diff
@@ -133,6 +133,14 @@
 		dismantle_wall(1)
 		return
 
+/turf/closed/wall/attack_basic_mob(mob/living/simple_animal/user, list/modifiers)
+	user.changeNext_move(CLICK_CD_MELEE)
+	user.do_attack_animation(src)
+	if((user.environment_smash & ENVIRONMENT_SMASH_WALLS) || (user.environment_smash & ENVIRONMENT_SMASH_RWALLS))
+		playsound(src, 'sound/effects/meteorimpact.ogg', 100, TRUE)
+		dismantle_wall(1)
```
Comment (Fikou) [inline code/game/turfs/closed/walls.dm:141]: true

File: code/game/turfs/closed/wall/reinf_walls.dm [hunk 3]
```diff
@@ -51,6 +51,19 @@
 		playsound(src, 'sound/effects/bang.ogg', 50, TRUE)
 		to_chat(user, span_warning("This wall is far too strong for you to destroy."))
 
+/turf/closed/wall/r_wall/attack_basic_mob(mob/living/simple_animal/user, list/modifiers)
+	user.changeNext_move(CLICK_CD_MELEE)
+	user.do_attack_animation(src)
+	if(!user.environment_smash)
+		return
+	if(user.environment_smash & ENVIRONMENT_SMASH_RWALLS)
+		dismantle_wall(1)
```
Comment (Fikou) [inline code/game/turfs/closed/wall/reinf_walls.dm:60]: true

**Comment by AMonkeyThatCodes** [comment]
this is ready

File: code/datums/ai/basic_mobs/basic_ai_behaviors/basic_destroy_obstacles.dm [hunk 4]
```diff
@@ -0,0 +1,40 @@
+//Smash obstacles between us and our target. Does not prevent replanning so can be cancelled out of nowhere and does not self-cancel.
+/datum/ai_behavior/basic_smash_obstacles
+	action_cooldown = 1 SECONDS
+	behavior_flags = AI_BEHAVIOR_MOVE_AND_PERFORM | AI_BEHAVIOR_ALLOWS_REPLANNING
+
+/datum/ai_behavior/basic_smash_obstacles/setup(datum/ai_controller/controller)
+	. = ..()
```
Comment (Watermelon914) [inline code/datums/ai/basic_mobs/basic_ai_behaviors/basic_destroy_obstacles.dm:7]: remove if you aren't using

File: code/datums/ai/basic_mobs/basic_ai_behaviors/basic_destroy_obstacles.dm [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+/datum/ai_behavior/basic_smash_obstacles/perform(delta_time, datum/ai_controller/controller)
+	. = ..()
+	var/mob/living/basic/mob_pawn = controller.pawn
+	if(mob_pawn.buckled)
+		mob_pawn.buckled.attack_basic_mob(src)
```
Comment (Watermelon914) [inline code/datums/ai/basic_mobs/basic_ai_behaviors/basic_destroy_obstacles.dm:13]:
```suggestion
	mob_pawn.buckled?.attack_basic_mob(src)
```

File: code/game/turfs/closed/walls.dm [hunk 6]
```diff
@@ -131,7 +131,15 @@
...
+		dismantle_wall(TRUE)
+		return
+
+/turf/closed/wall/attack_basic_mob(mob/living/simple_animal/user, list/modifiers)
+	user.changeNext_move(CLICK_CD_MELEE)
+	user.do_attack_animation(src)
+	if((user.environment_smash & ENVIRONMENT_SMASH_WALLS) || (user.environment_smash & ENVIRONMENT_SMASH_RWALLS))
+		playsound(src, 'sound/effects/meteorimpact.ogg', 100, TRUE)
+		dismantle_wall(TRUE)
 		return
```
Comment (Watermelon914) [inline code/game/turfs/closed/walls.dm:143]: don't need this return here

File: code/game/turfs/closed/walls.dm [hunk 7]
```diff
@@ -131,7 +131,15 @@
 	user.do_attack_animation(src)
 	if((user.environment_smash & ENVIRONMENT_SMASH_WALLS) || (user.environment_smash & ENVIRONMENT_SMASH_RWALLS))
 		playsound(src, 'sound/effects/meteorimpact.ogg', 100, TRUE)
-		dismantle_wall(1)
+		dismantle_wall(TRUE)
+		return
```
Comment (Watermelon914) [inline code/game/turfs/closed/walls.dm:135]: also don't need this return here

**Review by Watermelon914 — CHANGES_REQUESTED** [review]

**Comment by AMonkeyThatCodes** [comment]
cant talk to the other programmers on the discord

**Pull request closed without merging**